File: swift/aoc_day2/Sources/aoc_day2/Game.swift
import Foundation

protocol GameProtocol: Context {
  var gameId: UInt { get }
  var power: UInt { get }
  func isPossible(redCount: UInt, greenCount: UInt, blueCount: UInt) -> Bool
}

class Game: GameProtocol {
  var gameId: UInt = 0

  private(set) var lastColorContext: Color = .none

  private(set) var redCounts: [UInt] = []
  private(set) var greenCounts: [UInt] = []
  private(set) var blueCounts: [UInt] = []

  var power: UInt {
    (redCounts.max() ?? 0) * (greenCounts.max() ?? 0) * (blueCounts.max() ?? 0)
  }

  func isPossible(redCount: UInt, greenCount: UInt, blueCount: UInt) -> Bool {
    (redCounts.max() ?? 0) <= redCount &&
      (greenCounts.max() ?? 0) <= greenCount &&
      (blueCounts.max() ?? 0) <= blueCount
  }

  func setGameId(_ gameId: UInt) {
    self.gameId = gameId
  }

  func setColor(_ color: Color) {
    lastColorContext = color
  }

  func addCount(_ count: UInt) {
    switch lastColorContext {
    case .red:
      redCounts.append(count)
    case .green:
      greenCounts.append(count)
    case .blue:
      blueCounts.append(count)
    default:
      print("Invalid State")
    }
  }
}

File: 2023/c++/2023-2/src/Game.cpp
#include "Game.h"
#include <string>

using namespace std;

Game::Game(int id, int redCount, int greenCount, int blueCount) : id(id), redCount(redCount), blueCount(blueCount), greenCount(greenCount) { }

string Game::toString() {
  return "Game(" + to_string(id) + ", " + to_string(redCount) + ", " + to_string(greenCount) + ", " + to_string(blueCount) + ")";
}

bool Game::isPossible(int redCount, int greenCount, int blueCount) {
  return  true;
}


